src/base/bittorrent/torrenthandle.cpp

```diff
@@ -1744,45 +1749,64 @@ void TorrentHandle::handleFastResumeRejectedAlert(const lt::fastresume_rejected_
 
 void TorrentHandle::handleFileRenamedAlert(const lt::file_renamed_alert *p)
 {
-    const QString newName = Utils::Fs::fromNativePath(p->new_name());
-
-    // TODO: Check this!
-    if (filesCount() > 1) {
-        // Check if folders were renamed
-        QStringList oldPathParts = m_torrentInfo.origFilePath(p->index).split('/');
-        oldPathParts.removeLast();
-        QString oldPath = oldPathParts.join('/');
-        QStringList newPathParts = newName.split('/');
-        newPathParts.removeLast();
-        const QString newPath = newPathParts.join('/');
-        if (!newPathParts.isEmpty() && (oldPath != newPath)) {
-            qDebug("oldPath(%s) != newPath(%s)", qUtf8Printable(oldPath), qUtf8Printable(newPath));
-            oldPath = QString("%1/%2").arg(savePath(true), oldPath);
-            qDebug("Detected folder renaming, attempt to delete old folder: %s", qUtf8Printable(oldPath));
-            QDir().rmpath(oldPath);
-        }
-    }
-
     // We don't really need to call updateStatus() in this place.
     // All we need to do is make sure we have a valid instance of the TorrentInfo object.
     m_torrentInfo = TorrentInfo {m_nativeHandle.torrent_file()};
 
+    // remove empty leftover folders
+    const QString oldFilePath = m_oldPath[LTFileIndex {p->index}].takeFirst();
+    const QString newFilePath = Utils::Fs::fromNativePath(p->new_name());
+
+    if (m_oldPath[LTFileIndex {p->index}].isEmpty())
+        m_oldPath.remove(LTFileIndex {p->index});
+
+    QStringList oldPathParts = oldFilePath.split('/', QString::SkipEmptyParts);
+    oldPathParts.removeLast();  // drop file name part
+    QStringList newPathParts = newFilePath.split('/', QString::SkipEmptyParts);
+    newPathParts.removeLast();  // drop file name part
+
+#if defined(Q_OS_WIN)
+    const Qt::CaseSensitivity caseSensitivity = Qt::CaseInsensitive;
+#else
+    const Qt::CaseSensitivity caseSensitivity = Qt::CaseSensitive;
+#endif
+
+    int pathIdx = 0;
+    while ((pathIdx < oldPathParts.size()) && (pathIdx < newPathParts.size())) {
+        if (oldPathParts[pathIdx].compare(newPathParts[pathIdx], caseSensitivity) != 0)
+            break;
+        ++pathIdx;
+    }
+
+    for (int i = (oldPathParts.size() - 1); i >= pathIdx; --i) {
+        QDir().rmdir(savePath() + oldPathParts.join('/'));
+        oldPathParts.removeLast();
+    }
+
     --m_renameCount;
     while (!isMoveInProgress() && (m_renameCount == 0) && !m_moveFinishedTriggers.isEmpty())
         m_moveFinishedTriggers.takeFirst()();
+
+    if (isPaused() && (m_renameCount == 0))
+        saveResumeData();  // otherwise the new path will not be saved
 }
 
 void TorrentHandle::handleFileRenameFailedAlert(const lt::file_rename_failed_alert *p)
```

Comment:
I can not trigger this alert, even if I remove file or folder permissions, the file is not renamed, but this alert is not triggered.